### ISSUE TITLE
multiple mqtt_event_handler(), why?

### ISSUE THREAD

**Issue by Florida-Rob** [nopnop2002/esp-idf-mqtt-camera#9]
sorry if this is a noob question, but why do you have 2 x mqtt_event_handler() ? one in main.c the other main_sub.c, as both files included in the Makefile...
i dont see #if build flag, so confused
one uses 
xTaskNotifyGive( mqttBuf->taskHandle );
the other
xEventGroupClearBits(s_mqtt_event_group, MQTT_CONNECTED_BIT);

if the are both used, then totally confused...

I understand that to publish a message, then it has to be done in the CB, hence the xTaskNotify or xEventGroupClearBits, so another function can publish.  I really would have through that any function should be able to publish at any time, but understand its only within the CB when CONNECTED.
so shoudl we instead use esp_mqtt_client_enqueue(), so then the message just gets sent in the CB ?
or is that why you are using QueueHandle_t xQueueCmd;, si that publish messages are simply queued, then all sent together in the CB  notify with xTaskNotify  ?

**Comment by nopnop2002** [comment]
Please note that the mqtt_event_handler function is a static function.

mqtt_event_handler in main.c is used by app_main task, and mqtt_event_handler in mqtt_sub.c is used by mqtt_sub task.

Static functions can only be used from the same source file.

Therefore, mqtt_event_handler will not be multiple define.

**Comment by Florida-Rob** [comment]
yes, i see that, even with the same function call, esp_mqtt_client_register_event(), with the same variables. so i would have thought registering to the same event..., same handler... mqtt_event_handler() twice !
i was confused by the esp-idf example where the handler was called in a function from main (not within main), from mqtt_app_start()

so in my example need (for other functions to publish to MQTT), i should use your mqtt_sub() as an example, and post commands to the mqttBuf, which are then sent with mqtt_sub() and handler..
i see you 'subscribe' on CONNECTED, i suppose equally you can 'publish' here ?
would be really helpful for a write-up for how this all ties together....

**Comment by nopnop2002** [comment]
>i see you 'subscribe' on CONNECTED, i suppose equally you can 'publish' here ?

It is possible.

**Comment by Florida-Rob** [comment]
do you have a write-up how your code is structured ?
the flow of entry of commands into the buffer cmdBuf?
and who posts to MQTT, and when ?

so, the flow is:
1) mqtt_sub(), if subscribe message xTaskNotifyTake (from mqtt_event_handler, with xTaskNotifyGive), send with xQueueSend cmdBuf
2) app_main(), waiting xQueueReceive cmdBuf, wait for MQTT_CONNECTED_BIT, test for *ANY*  command but CMD_HALT, then publish image

**Comment by Florida-Rob** [comment]
p.s by trial and error, i agree with "A handler is required for each task."
i don't know why...
if you have a main with the handler, then a sub task.  even if you set a global variable, or xTaskNotify/xTaskNotifyWait, it will always panic and crash ! you must use a "A handler is required for each task."

**Comment by nopnop2002** [comment]
```
so, the flow is:

mqtt_sub(), if subscribe message xTaskNotifyTake (from mqtt_event_handler, with xTaskNotifyGive), send with xQueueSend cmdBuf
app_main(), waiting xQueueReceive cmdBuf, wait for MQTT_CONNECTED_BIT, test for ANY command but CMD_HALT, then publish image
``` 

You are right.

>p.s by trial and error, i agree with "A handler is required for each task."

The address of the MQTT handler function is registered with esp_mqtt_client_register_event().
But if the handler callback function does not exist on the same source, I think this is because the address cannot be resolved or it becomes a FAR jump.


The MQTT API can be found here, but there is no detailed explanation about handler callbacks.

https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/protocols/mqtt.html
